File: norcal/Coder.c
#define _CRT_SECURE_NO_WARNINGS
#include <assert.h>
#include <stdio.h>
#include "Common.h"

// Documented at http://wiki.nesdev.com/w/index.php/INES
static uint8_t Header[16] =
{
    0x4E, 0x45, 0x53, 0x1A,     // iNES file signature
    0x02,                       // PRG ROM size
    0x01,                       // CHR ROM size
    0x01, 0x00, 0x00, 0x00,     // various flags
    0x00, 0x00, 0x00, 0x00,     // unused
    0x00, 0x00,
};

#define PrgRomBase 0x8000

static uint8_t Prg[32 * 1024];
static uint8_t Chr[8 * 1024];
static int32_t PrgNext;

static FILE *CommentFile;

static void Append(uint8_t b)
{
    if (PrgNext >= sizeof(Prg)) Error("program size limit exceeded");
    Prg[PrgNext++] = b;
}

void Emit(Opcode op)
{
    Append(op);
}

void Emit_U8(Opcode op, uint8_t arg)
{
    Append(op);
    Append(arg);
}

void Emit_U16(Opcode op, uint16_t arg)
{
    Append(op);
    Append((uint8_t)arg);
    Append((uint8_t)(arg >> 8));
}

void EmitComment(char *comment)
{
    if (!CommentFile)
    {
        // TODO: Choose a more appropriate path for the comments file.
        CommentFile = fopen("comments.txt", "w");
        assert(CommentFile);
    }

    char line[256];
    int len = sprintf(line, "%04X    %s\n", (uint16_t)PrgNext, comment);
    assert(fwrite(line, len, 1, CommentFile) == 1);
    fflush(CommentFile);
}

void EmitFix_S8(int32_t address, int32_t target)
{
    int32_t romOffset = address - PrgRomBase;
    if (romOffset < 0 || romOffset >= PrgNext) Panic("fixup address is out of bounds");
    int32_t offset = target - address - 1;
    if (offset < INT8_MIN || offset > INT8_MAX) Panic("branch distance is too far");
    Prg[romOffset] = (uint8_t)offset;
}

void EmitFix_U16(int32_t address, int32_t target)
{
    int32_t romOffset = address - PrgRomBase;
    if (romOffset < 0 || romOffset + 1 >= PrgNext) Panic("fixup address is out of bounds");
    Prg[romOffset] = LowByte(target);
    Prg[romOffset + 1] = HighByte(target);
}

int32_t GetCurrentCodeAddress()
{
    return PrgRomBase + PrgNext;
}

void WriteImage(char *filename)
{
    // HACK: Hardcoded interrupt vector:
    // Interrupts: nmi, reset, irq
    Prg[0x7FFA] = 0x00;
    Prg[0x7FFB] = 0x80;
    Prg[0x7FFC] = 0x00;
    Prg[0x7FFD] = 0x80;
    Prg[0x7FFE] = 0x00;
    Prg[0x7FFF] = 0x80;

    // TODO: Use a specified CHR ROM input file.

    FILE *f = fopen(filename, "wb");
    assert(f);
    assert(fwrite(Header, sizeof(Header), 1, f) == 1);
    assert(fwrite(Prg, sizeof(Prg), 1, f) == 1);
    assert(fwrite(Chr, sizeof(Chr), 1, f) == 1);
    fclose(f);
}

File: norcal/Coder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

partial class Compiler
{
    List<Expr> Assembly = new List<Expr>();

    // PRG must be exactly 32k, and CHR must be exactly 8k.
    static readonly int ChrRomSize = 0x2000;
    static readonly int PrgRomBase = 0x8000;
    static readonly int PrgRomSize = 0x8000;

    // Documented at http://wiki.nesdev.com/w/index.php/INES
    static readonly byte[] Header = new byte[16]
    {
        0x4E, 0x45, 0x53, 0x1A,     // iNES file signature
        0x02,                       // PRG ROM size
        0x01,                       // CHR ROM size
        0x01, 0x00, 0x00, 0x00,     // various flags
        0x00, 0x00, 0x00, 0x00,     // unused
        0x00, 0x00,
    };

    void Emit(params object[] args)
    {
        Expr e = Expr.Make(args);
        Assembly.Add(e);
    }

    void EmitComment(string message)
    {
        Emit(Asm.Comment, message);
    }

    public void Assemble(string outputFilename)
    {
        // TODO: Use a specified CHR ROM input file.
        byte[] chr = new byte[ChrRomSize];

        // First pass: Calculate label values.
        Dictionary<string, int> labels = new Dictionary<string, int>();
        RunPass(
            Assembly,
            (label, address) =>
            {
                // TODO: Make sure the address is in the allowed range.
                labels[label] = address;
            },
            (label, baseAddress) => 0);

        // Second pass: Generate machine code.
        List<byte> prg = RunPass(
            Assembly,
            (label, address) => { /* ignored */ },
            (label, baseAddress) =>
            {
                int address;
                if (labels.TryGetValue(label, out address)) return address - baseAddress;
                Program.Panic("assembler: label not defined: {0}", label);
                return 0;
            });

        // TODO: Make sure that the PRG ROM size limit isn't exceeded. It must not overwrite the vector table.
        if (prg.Count != PrgRomSize) Program.Error("assembler: program has wrong length");

        using (BinaryWriter w = new BinaryWriter(new FileStream(outputFilename, FileMode.Create)))
        {
            w.Write(Header);
            w.Write(prg.ToArray());
            w.Write(chr);
        }
    }

    enum LabelType
    {
        Relative,
        Absolute,
    }

    static List<byte> RunPass(List<Expr> input, Action<string, int> defineLabel, Func<string, int, int> findLabel)
    {
        List<byte> prg = new List<byte>();

        foreach (Expr e in input)
        {
            string label;
            int skipTarget;

            if (e.MatchTag(Asm.Comment))
            {
                // Ignore.
            }
            else if (e.Match(Asm.Function, out label) || e.Match(Asm.Label, out label))
            {
                defineLabel(label, PrgRomBase + prg.Count);
            }
            else if (e.Match(Asm.SkipTo, out skipTarget))
            {
                skipTarget -= PrgRomBase;

                // Pad out the code until reaching the target address:
                if (skipTarget > ushort.MaxValue) Program.Panic("assembler: skip address is too large");
                if (prg.Count > skipTarget) Program.Panic("assembler: cannot skip backward");
                while (prg.Count < skipTarget)
                {
                    prg.Add(0);
                }
            }
            else if (e.Match(Asm.Word, out label))
            {
                int address = findLabel(label, 0);
                prg.Add(LowByte(address));
                prg.Add(HighByte(address));
            }
            else
            {
                string mnemonic = "???";
                int operand = 0;
                string modifier;
                int operandFormat = AsmInfo.INV;
                if (e.Match(out mnemonic))
                {
                    operandFormat = AsmInfo.IMP;
                }
                else if (e.Match(out mnemonic, out operand, out modifier))
                {
                    if (modifier == Asm.Absolute) operandFormat = AsmInfo.ABS;
                    else if (modifier == Asm.Immediate) operandFormat = AsmInfo.IMM;
                    else if (modifier == Asm.IndirectY) operandFormat = AsmInfo.ZYI;
                    else Program.Panic("unknown assembly modifier: {0}", modifier);
                }
                else if (e.Match(out mnemonic, out label))
                {
                    int baseAddress;
                    if (AsmInfo.ShortJumpInstructions.Contains(mnemonic))
                    {
                        operandFormat = AsmInfo.REL;
                        // Relative jumps are always relative to the address of the *following* instruction:
                        baseAddress = PrgRomBase + prg.Count + 2;
                    }
                    else if (AsmInfo.LongJumpInstructions.Contains(mnemonic))
                    {
                        operandFormat = AsmInfo.ABS;
                        baseAddress = 0;
                    }
                    else
                    {
                        operandFormat = AsmInfo.INV;
                        baseAddress = 0;
                    }
                    operand = findLabel(label, baseAddress);
                }
                else
                {
                    Program.Panic("invalid instruction format: {0}", e.Show());
                }

                int formalSize = AsmInfo.OperandSizes[operandFormat];

                int actualSize;
                if (operandFormat == AsmInfo.IMP) actualSize = 0;
                else if (operand >= sbyte.MinValue && operand <= byte.MaxValue) actualSize = 1;
                else if (operand >= short.MinValue && operand <= ushort.MaxValue) actualSize = 2;
                else actualSize = 99;

                if (actualSize > formalSize)
                {
                    Program.Panic("assembly operand is too large: {0}", e.Show());
                }

                // Search for a matching instruction:
                List<byte> candidates = new List<byte>();
                for (int opcode = 0; opcode < 256; opcode++)
                {
                    if (mnemonic == AsmInfo.Mnemonics[opcode] &&
                        operandFormat == AsmInfo.OperandFormats[opcode])
                    {
                        candidates.Add((byte)opcode);
                    }
                }

                if (candidates.Count == 1)
                {
                    if (AsmInfo.ShortJumpInstructions.Contains(mnemonic) && (operand < sbyte.MinValue || operand > sbyte.MaxValue))
                    {
                        Program.Panic("relative branch offset is too large");
                    }

                    prg.Add(candidates[0]);

                    if (formalSize == 1)
                    {
                        prg.Add(LowByte(operand));
                    }
                    else if (formalSize == 2)
                    {
                        prg.Add(LowByte(operand));
                        prg.Add(HighByte(operand));
                    }
                }
                else
                {
                    Program.Panic("invalid or ambiguous instruction: {0}", e.Show());
                }
            }
        }

        return prg;
    }
}

public class Instruction
{
    public byte Opcode;
    public int Operand;
    public int OperandFormat;
}

public static class Asm
{
    // Directives:
    public static readonly string Comment = "comment";
    public static readonly string Function = "function";
    public static readonly string Label = "label";
    public static readonly string SkipTo = "skip_to";
    public static readonly string Word = "word";

    // Instruction modifiers:
    public static readonly string Absolute = "absolute";
    public static readonly string Immediate = "immediate";
    public static readonly string IndirectY = "indirect_y";
}
